File: libraries/javalib/java/lang/reflect/Constructor.java
/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998, 2001
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 *
 * Checked Spec: JDK 1.3
 */

package java.lang.reflect;


public final class Constructor
  extends AccessibleObject
  implements Member
{
	private Class clazz;
	private int slot;
	private Class[] parameterTypes;
	private Class[] exceptionTypes;

private Constructor() {
}

public boolean equals(Object obj)
	{
	// Catch the simple case where they're really the same
	if (this == obj) {
		return (true);
	}

	// if obj is null or not Constructor then they are not the same
	if (!(obj instanceof Constructor)) {
		return (false);
	}

	Constructor cobj = (Constructor)obj;

	if (clazz != cobj.clazz) {
		return (false);
	}
	if (parameterTypes.length != cobj.parameterTypes.length) {
		return (false);
	}
	for (int i = parameterTypes.length; i-- > 0; ) {
		if (parameterTypes[i] != cobj.parameterTypes[i]) {
			return (false);
		}
	}
	return (true);
}

public Class getDeclaringClass()
	{
	return (clazz);
}

public Class[] getExceptionTypes()
	{
	return (exceptionTypes);
}

native public int getModifiers();

public String getName()
	{
	return (clazz.getName());
}

public Class[] getParameterTypes()
	{
	return (parameterTypes);
}

public int hashCode()
	{
	return (clazz.getName().hashCode());
}

public Object newInstance(Object initargs[]) throws InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
	Method meth = new Method(clazz,
				 slot,
				 "<init>",
				 clazz,
				 parameterTypes,
				 exceptionTypes);
	return (meth.invoke(null, initargs));
}

public String toString()
	{
	StringBuffer str = new StringBuffer();
	int mod = getModifiers();

	// Modifier
	if (mod != 0) {
		str.append(Modifier.toString(mod));
		str.append(' ');
	}

	// Class name.
	str.append(clazz.getName());
	str.append('(');

	// Signature
	for (int i = 0; i < parameterTypes.length; i++) {
		str.append(Method.getPrettyName(parameterTypes[i]));
		if (i+1 < parameterTypes.length) {
			str.append(',');
		}
	}
	str.append(')');

        if (exceptionTypes.length > 0) {
	        str.append(" throws ");
                for (int i = 0; i < exceptionTypes.length; i++) {
                        str.append(exceptionTypes[i].getName());
                        if (i+1 < exceptionTypes.length) {
			        str.append(',');
                        }
                }
        }

	return (str.toString());
}
}

File: libraries/clib/native/Constructor.c
/*
 * java.lang.reflect.Constructor.c
 *
 * Copyright (c) 1996, 1997
 *	Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "lib-license.terms" for information on usage and redistribution 
 * of this file. 
 */

#include "config.h"
#include "config-std.h"
#include "config-mem.h"
#include "../../../kaffe/kaffevm/gtypes.h"
#include "../../../kaffe/kaffevm/access.h"
#include "../../../kaffe/kaffevm/constants.h"
#include "../../../kaffe/kaffevm/object.h"
#include "../../../kaffe/kaffevm/classMethod.h"
#include "../../../kaffe/kaffevm/itypes.h"
#include "../../../kaffe/kaffevm/support.h"
#include "../../../kaffe/kaffevm/baseClasses.h"
#include "Constructor.h"
#include "Method.h"
#include <native.h>
#include <jni.h>
#include "defs.h"

/* XXX move in header file */
jobject
Java_java_lang_reflect_Method_invoke(JNIEnv* env, jobject _this, 
	jobject _obj, jarray _argobj);

jint
java_lang_reflect_Constructor_getModifiers(struct Hjava_lang_reflect_Constructor* this)
{
	Hjava_lang_Class* clazz;
	jint slot;

	clazz = unhand(this)->clazz;
	slot = unhand(this)->slot;

	assert(slot < clazz->nmethods);

	return (clazz->methods[slot].accflags & ACC_MASK);
}

jobject
Java_java_lang_reflect_Constructor_newInstance(JNIEnv* env, jobject _this, jarray argobj)
{
	/* 
	 * We fake a java.lang.reflect.Method Object so that we can call
	 * Java_java_lang_reflect_Method_invoke from here.
	 *
	 * The main task Method_invoke does is converting the jarray of
	 * arguments into a jvalue * array.  Method.invoke will invoke the
	 * right JNI function when it detects that the method is a constructor
	 */
	Hjava_lang_reflect_Method meth[1];

	Hjava_lang_reflect_Constructor* this = (Hjava_lang_reflect_Constructor*)_this;

	unhand(meth)->clazz = unhand(this)->clazz;
	unhand(meth)->slot = unhand(this)->slot;
	unhand(meth)->name = 0; /* It's a constructor!! */
	unhand(meth)->returnType = unhand(this)->clazz;
	unhand(meth)->parameterTypes = unhand(this)->parameterTypes;
	unhand(meth)->exceptionTypes = unhand(this)->exceptionTypes;

	return Java_java_lang_reflect_Method_invoke(env, (jobject)meth, 
			(jobject)NULL, argobj);
}
